src/axom/quest/MarchingCubes.hpp

```diff
@@ -0,0 +1,342 @@
+// Copyright (c) 2017-2023, Lawrence Livermore National Security, LLC and
+// other Axom Project Developers. See the top-level LICENSE file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+
+/*!
+ * \file MarchingCubes.hpp
+ *
+ * \brief Consists of class implementing marching cubes algorithm to
+ * compute iso-surface from a scalar field in a blueprint mesh.
+ */
+
+#ifndef AXOM_PRIMAL_MARCHINGCUBES_H_
+#define AXOM_PRIMAL_MARCHINGCUBES_H_
+
+// Axom includes
+#include "axom/config.hpp"
+#include "axom/mint/mesh/UnstructuredMesh.hpp"
+
+// Conduit includes
+#include "conduit_node.hpp"
+
+// C++ includes
+#include <string>
+
+// Add some helper preprocessor defines for using OPENMP, CUDA, and HIP policies
+// within the marching cubes implementation.
+#if defined(AXOM_USE_RAJA)
+  #ifdef AXOM_USE_OPENMP
+    #define _AXOM_MC_USE_OPENMP
+  #endif
+  #if defined(AXOM_USE_CUDA) && defined(AXOM_USE_UMPIRE)
+    #define _AXOM_MC_USE_CUDA
+  #endif
+  #if defined(AXOM_USE_HIP) && defined(AXOM_USE_UMPIRE)
+    #define _AXOM_MC_USE_HIP
+  #endif
+#endif
+
+namespace axom
+{
+namespace quest
+{
+/*!
+  @brief Enum for runtime execution policy
+
+  The policy implicitly selects the execution space and allocator id.
+*/
+enum class MarchingCubesRuntimePolicy
+{
+  seq = 0,
+  omp = 1,
+  cuda = 2,
+  hip = 3
+};
+
+/// Utility function to allow formating a MarchingCubesRuntimePolicy
+inline auto format_as(MarchingCubesRuntimePolicy pol)
+{
+  return fmt::underlying(pol);
+}
+
+/// Predicate to determine if a given \a MarchingCubesRuntimePolicy is valid for this configuration
+inline bool isValidRuntimePolicy(MarchingCubesRuntimePolicy policy)
+{
+  switch(policy)
+  {
+  case MarchingCubesRuntimePolicy::seq:
+    return true;
+
+  case MarchingCubesRuntimePolicy::omp:
+#ifdef _AXOM_MC_USE_OPENMP
+    return true;
+#else
+    return false;
+#endif
+
+  case MarchingCubesRuntimePolicy::cuda:
+#ifdef _AXOM_MC_USE_CUDA
+    return true;
+#else
+    return false;
+#endif
+  case MarchingCubesRuntimePolicy::hip:
+#ifdef _AXOM_MC_USE_HIP
+    return true;
+#else
+    return false;
+#endif
+  }
+
+  return false;
+}
+
+/*!
+ * \@brief Class implementing marching cubes algorithm on a single
+ * structured mesh domain, alone or within a multi-domain mesh.
+ *
+ * Implementation is for 2D (marching squares) and 3D (marching
+ * cubes).
+ *
+ * The input mesh is a Conduit::Node following the Mesh Blueprint
+ * convention.
+ *
+ * Usage example:
+ * @beginverbatim
+ *   MarchingCubes computationalMesh(meshNode, "coords");
+ *   double contourValue = 0.0;
+ *   set_function_field("my_function");
```

Comment:
```suggestion
 *   mc.set_function_field("my_function");
```
I think that `set_function_field` is not a free function, but a method of class `MarchingCubes`.